client/src/plugins/usage-statistics/event-handlers/DeploymentEventHandler.js

```diff
@@ -0,0 +1,96 @@
+/**
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
+ * under one or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information regarding copyright
+ * ownership.
+ *
+ * Camunda licenses this file to you under the MIT; you may not use this file
+ * except in compliance with the MIT License.
+ */
+
+import BaseEventHandler from './BaseEventHandler';
+
+import { extractProcessVariables } from '../../../util';
+
+const RELEVANT_TAB_TYPES = ['bpmn', 'dmn'];
+
+// Sends a deployment event to ET everytime when a user triggers a deploy to
+// the Camunda Engine, ignoring cmmn deployments
+export default class DeploymentEventHandler extends BaseEventHandler {
+
+  constructor(params) {
+
+    const { onSend, subscribe } = params;
+
+    super('deployment', onSend);
+
+    subscribe('deployment.done', this.handleDeployment);
+    subscribe('deployment.error', this.handleDeployment);
+  }
+
+  generateMetrics = async (file, tabType) => {
+    let metrics = {};
+
+    // (1) process variables (bpmn only)
+    if (tabType === 'bpmn' && file.contents) {
+      const processVariables = await extractProcessVariables(file);
+
+      metrics = {
+        processVariablesCount: processVariables.length,
+        ...metrics
+      };
+    }
+
+    return metrics;
+  }
+
+  handleDeployment = async (context) => {
+    const {
+      error,
+      tab,
+      isRun
+    } = context;
+
+    const {
+      type,
+      file
+    } = tab;
+
+    // (0) check whether usage statistics are enabled
+    if (!this.isEnabled()) {
+      return;
+    }
+
+    // (1) ensure bpmn or dmn deployments
+    if (!RELEVANT_TAB_TYPES.includes(type)) {
+      return;
+    }
+
+    // (2) retrieve deployment status
+    const outcome = error ? 'failure' : 'success';
+
+    // (3) generate diagram related metrics, e.g. process variables
+    const diagramMetrics = await this.generateMetrics(file, type);
+
+    let payload = {
+      diagramType: type,
+      deployment: {
+        outcome
+      },
+      diagramMetrics
+    };
+
+    // (4) retrieve deployment error
+    if (error) {
+      payload.deployment.error = error.code;
+    }
+
+    // (5) indicate deployments triggered via RUN
+    if (isRun) {
```

Comment:
Would allow us to get rid of this thing, too which I personally think is a code smell, too.

Reply:
👍